File: ipc/receiver.lua
local semlib = require("ipc.sem")
local shmlib = require("ipc.shm")
local ffi = require("ffi")
ffi.cdef[[
struct shm_data_t {
	char data[255][1024];
	int capacity;
	int length;
	int start; // read index
	int end_idx; // write index
};
]]

local RING_LEN = 255

local sem_send = assert(semlib.open("sem_send", RING_LEN))
local sem_recv = assert(semlib.open("sem_recv", 1))
local shm = assert(shmlib.create("myshm", 261136))
local sem_mutex = assert(semlib.open("sem_mutex", 1))

-- need to decrement here because semlib.open can't create semaphore with initial value = 0
sem_recv:dec()
local recv = 0
local shm_data = ffi.cast("struct shm_data_t *", shm:addr())
while recv < 10000 do
	sem_recv:dec()
	sem_mutex:dec()

	-- use the data memcpy(data, shm_data->data[shm_data->start], 1024);
	shm_data.start = (shm_data.start + 1) % RING_LEN;
	shm_data.length = shm_data.length -1;
	recv = recv + 1

	--print("recv=", recv)
	
	sem_mutex:inc()
	sem_send:inc()
end

File: ipc/receiver.c
#include <stdio.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <unistd.h>
#include <fcntl.h>
#include <semaphore.h>
#include <sys/stat.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>

#define RING_LEN 255

struct shm_data_t {
	char data[RING_LEN][1024];
	int capacity;
	int length;
	int start; // read index
	int end_idx; // write index
};
#define SEM_NAME "mysem"
#define SEM_MUTEX "sem_mutex"

int main() {
	int shm_fd;
	struct shm_data_t *shm_data;
	int shm_size = sizeof(struct shm_data_t);


	// create semaphore
	sem_t *sem_id =sem_open(SEM_NAME, O_CREAT, S_IRUSR | S_IWUSR, 0);
	sem_t *sem_mutex = sem_open(SEM_MUTEX, O_CREAT, S_IRUSR | S_IWUSR, 1);

	// create shared memory
	shm_fd = shm_open("myshm",  O_CREAT | O_RDWR, S_IRWXU | S_IRWXG);
	if (shm_fd < 0) {
		perror("in shmopen");
		exit(1);
	}

	printf("shm_size=%d\n", shm_size);

	ftruncate(shm_fd, shm_size);

	// map the memory
	shm_data = (struct shm_data_t *)mmap(NULL, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
	if (shm_data == NULL) {
		perror("mmap failed");
		exit(1);
	}

	// write data
	int i = 0;
	shm_data->capacity = RING_LEN;
	shm_data->length = 0;
	char data[1024];
	for (i = 0; i < 10000; i++) {
		sem_wait(sem_id);
		sem_wait(sem_mutex);
		
		memcpy(data, shm_data->data[shm_data->start], 1024);
		shm_data->start = (shm_data->start + 1) % RING_LEN;
		shm_data->length = shm_data->length -1;
		
		sem_post(sem_mutex);
	}
	printf("got 10000 data\n");
	fflush(stdout);
	
	if (shm_unlink("/myshm") != 0) {
		perror("shmunlink failed");
		exit(1);
	}

	if (sem_unlink(SEM_NAME) != 0) {
		perror("sem_unlink failed");
		exit(1);
	}
	if (sem_unlink(SEM_MUTEX) != 0) {
		perror("sem_unlink failed");
		exit(1);
	}
	return 0;
}
